cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1196 @@
+			zap.Uint64("regionID", req.RegionId),
+			zap.Uint64("storeID", rs.storeID),
+			zap.String("addr", rs.storeAddr),
+			zap.Uint64("streamID", stream.streamID))
+		return nil
+	}
+
+	fetchMoreReq := func() (sri singleRegionInfo, _ error) {
+		waitReqTicker := time.NewTicker(60 * time.Second)
+		defer waitReqTicker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return sri, ctx.Err()
+			case sri = <-stream.requests.Out():
+				return sri, nil
+			case <-waitReqTicker.C:
+				if stream.countStates() == 0 {
+					return sri, errors.New("closed as idle")
+				}
+			}
+		}
+	}
+
+	sri := *stream.preFetchForConnecting
+	stream.preFetchForConnecting = nil
+	for {
+		if sri.lockedRange == nil {
+			// It means it's a special task from stopping the table.
+			req := &cdcpb.ChangeDataRequest{
+				RequestId: uint64(sri.requestedTable.subscriptionID),
+				Request:   &cdcpb.ChangeDataRequest_Deregister_{},
+			}
+			if err = doSend(req); err != nil {
+				return err
+			}
+			// NOTE: some principles to help understand deregistering a table:
+			// 1. after a Deregister(requestID) message is sent out, no more region requests
+			//    with the same requestID will be sent out in the same GRPC stream;
+			// 2. so it's OK to clear all pending states in the GRPC stream;
+			// 3. is it possible that TiKV is keeping to send events belong to a removed state?
+			//    I guess no because internal errors will cause the changefeed or table stopped,
+			//    and then those regions from the bad requestID will be unsubscribed finally.
+			for _, state := range stream.takeStates(sri.requestedTable.subscriptionID) {
+				state.markStopped(&sendRequestToStoreErr{})
+				slot := hashRegionID(state.sri.verID.GetID(), len(s.workers))
+				sfEvent := statefulEvent{eventItem: eventItem{state: state}}
+				if err = s.workers[slot].sendEvent(ctx, sfEvent); err != nil {
+					return errors.Trace(err)
+				}
+			}
+		} else if sri.requestedTable.stopped.Load() {
+			// It can be skipped directly because there must be no pending states from
+			// the stopped requestedTable, or the special singleRegionInfo for stopping
+			// the table will be handled later.
+			s.onRegionFail(newRegionErrorInfo(sri, &sendRequestToStoreErr{}))
+		} else {
+			connectTime := time.Since(sri.lockedRange.Created).Milliseconds()
+			s.metrics.regionConnectDuration.Observe(float64(connectTime))
+
+			subscriptionID := sri.requestedTable.subscriptionID
+			state := newRegionFeedState(sri, uint64(subscriptionID))
+			state.start()
+			stream.setState(subscriptionID, sri.verID.GetID(), state)
+
+			req := s.createRegionRequest(sri)
+			if err = doSend(req); err != nil {
+				return err
+			}
+		}
+
+		if sri, err = fetchMoreReq(); err != nil {
+			return err
+		}
+	}
+}
+
+func (s *SharedClient) receiveFromStream(ctx context.Context, rs *requestedStore, stream *requestedStream) (err error) {
+	for {
+		cevent, err := stream.client.Recv()
+		if err != nil {
+			code := grpc.Code(err)
```

Comment:
<img width="842" alt="image" src="https://github.com/pingcap/tiflow/assets/61726649/c3a4a03d-1433-483b-87fc-b6a38db659bd">
grpc.Code is deprecated: use status.Code instead.